### PR TITLE
Flowc cpp3 branch 1

### PR DESCRIPTION
**Pull request by dmitry-vlasov**

### CODE REVIEW

File: tools/flowc/incremental/fi_helpers.flow [hunk 1]
```diff
@@ -367,7 +378,7 @@ fiMatchTypars(type : FiType, pattern : FiType, acc : Tree<string, FiType>) -> Tr
 			}
 		FiTypeRef(t): 
 			switch (pattern) {
-				FiTypeArray(tp): fiMatchTypars(t, tp, acc);
+				FiTypeRef(tp): fiMatchTypars(t, tp, acc);
```
Comment (alstrup) [inline tools/flowc/incremental/fi_helpers.flow:381]: Looks like a general bug fix

File: tools/flowc/statements/fsexp.flow [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+		FsAssign(var: FsVar, e: FsExp);
+
+	FsExp ::= FsLambda, FsCall, FsCallPrim, FsTernary, FsCast, FsRequire, FsUnsafe, FsVar, FiConst;
+		FsLambda(closure: [string], args: [FiFunArg], body: FsStat, type : FiTypeFunction);
+		FsCall(f: FsExp, args: [FsExp], type : FiType);
+		FsCallPrim(op: FcPrim, es: [FsExp], type: FiType);
+		FsRequire(flowfile : string, e : FsExp, type : FiType);
+		FsUnsafe(name : string, fallback : FsExp, type : FiType);
+		FsCast(e: FsExp, from : FiType, to : FiType);
+		FsTernary(cond: FsExp, s1: FsExp, s2: FsExp, type: FiType);
```
Comment (alstrup) [inline tools/flowc/statements/fsexp.flow:24]: TernaryIf would be better.

File: tools/flowc/statements/fsexp.flow [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+export {
+	FsStat ::= FsLet, FsIf, FsSwitch, FsSeq, FsWrap, FsRet, FsTailCall, FsVarDecl, FsAssign;
+
+		FsLet(name: string, vtype: FiType, e: FsExp);
+		FsIf(cond: FsExp, s1: FsStat, s2: FsStat, type: FiType);
+		FsSwitch(x: FsVar, switchType: FiType, cases: [FsCase], type: FiType);
+			FsCase(struct: string, args: [string], body: FsStat);
+		FsSeq(ss: [FsStat], type: FiType);
+		FsWrap(e: FsExp);
```
Comment (alstrup) [inline tools/flowc/statements/fsexp.flow:11]: Not sure what this one is. Add comment?

Comment (alstrup) [reply tools/flowc/statements/fsexp.flow:11]: Oh, I see. It is to lift an expression to a statement. Then WrapExp is better

File: tools/flowc/incremental/fi_helpers.flow (hunk 1 above)
Comment (dmitry-vlasov) [reply tools/flowc/incremental/fi_helpers.flow:381]: Yes, it's true. It was a bug.

File: tools/flowc/statements/fsexp.flow (hunk 2 above)
Comment (dmitry-vlasov) [reply tools/flowc/statements/fsexp.flow:24]: Done

**Pull request merged**